cache/clock_cache.h

```diff
@@ -27,132 +29,268 @@ namespace ROCKSDB_NAMESPACE {
-    refs = 0;
-    flags = 0;
-    SetIsVisible(false);
+  // The number of elements that hash to this slot or a lower one, but wind
+  // up in this slot or a higher one.
+  std::atomic<uint32_t> displacements;
+
+  // Synchronization rules:
+  // - Use a shared reference when we want the handle's identity
+  //    members (key_data, hash, value and IS_ELEMENT flag) to
+  //    remain untouched, but not modify them. The only updates
+  //    that a shared reference allows are:
+  //      * set CLOCK_PRIORITY to NONE;
+  //      * set the HAS_HIT bit.
+  //    Notice that these two types of updates are idempotent, so
+  //    they don't require synchronization across shared references.
+  // - Use an exclusive reference when we want identity members
+  //    to remain untouched, as well as modify any identity member
+  //    or flag.
+  // - displacements can be modified without holding a reference.
+  // - refs is only modified through appropiate functions to
+  //    take or release references.
+
+  ClockHandle()
+      : value(nullptr),
+        deleter(nullptr),
+        hash(0),
+        total_charge(0),
+        refs(0),
+        flags(0),
+        displacements(0) {
+    SetWillBeDeleted(false);
     SetIsElement(false);
     SetClockPriority(ClockPriority::NONE);
     SetCachePriority(Cache::Priority::LOW);
-    displacements = 0;
     key_data.fill(0);
   }
 
-  Slice key() const { return Slice(key_data.data(), kCacheKeySize); }
-
-  // Increase the reference count by 1.
-  void Ref() { refs++; }
-
-  // Just reduce the reference count by 1. Return true if it was last reference.
-  bool Unref() {
-    assert(refs > 0);
-    refs--;
-    return refs == 0;
+  ClockHandle(const ClockHandle& other) { *this = other; }
+
+  void operator=(const ClockHandle& other) {
+    value = other.value;
+    deleter = other.deleter;
+    hash = other.hash;
+    total_charge = other.total_charge;
+    refs.store(other.refs);
+    key_data = other.key_data;
+    flags.store(other.flags);
+    SetWillBeDeleted(other.WillBeDeleted());
+    SetIsElement(other.IsElement());
+    SetClockPriority(other.GetClockPriority());
+    SetCachePriority(other.GetCachePriority());
+    displacements.store(other.displacements);
   }
 
-  // Return true if there are external refs, false otherwise.
-  bool HasRefs() const { return refs > 0; }
+  Slice key() const { return Slice(key_data.data(), kCacheKeySize); }
 
-  bool IsVisible() const { return flags & IS_VISIBLE; }
-
-  void SetIsVisible(bool is_visible) {
-    if (is_visible) {
-      flags |= IS_VISIBLE;
-    } else {
-      flags &= ~IS_VISIBLE;
-    }
-  }
+  bool HasExternalRefs() const { return (refs & EXTERNAL_REFS) > 0; }
 
   bool IsElement() const { return flags & IS_ELEMENT; }
 
   void SetIsElement(bool is_element) {
     if (is_element) {
       flags |= IS_ELEMENT;
     } else {
-      flags &= ~IS_ELEMENT;
+      // flags &= static_cast<uint8_t>(~IS_ELEMENT);
+      flags.fetch_and(~IS_ELEMENT);
```

Comment:
This doesn't work. Thanks for the try!

Reply:
The problem is that when an arithmetic (or bitwise) operator is applied to a type smaller than int, it is promoted to int. Some compilers aren't forgiving even when you can tell from the input types that there's no data loss.

Reply:
(so you need to static_cast back to unsigned char)